internal/framework/subproviders/morpheus/resources/datastore/datastore_read.go

```diff
@@ -321,6 +353,249 @@ func getDatastoreAsState(
+	if err != nil {
+		diags.AddWarning(
+			"populate datastore resource",
+			fmt.Sprintf("datastore %d cloud datastore GET failed: %s", id, errors.ErrMsg(err, hresp)),
+		)
+
+		return types.SetNull(TenantsType{}), NewResourcePermissionsValueNull(), diags
+	}
+
+	cloudDatastore, ok := cdResp.GetDatastoreOk()
+	if !ok || cloudDatastore == nil {
+		diags.AddWarning(
+			"populate datastore resource",
+			fmt.Sprintf("datastore %d missing cloud datastore in response", id),
+		)
+
+		return types.SetNull(TenantsType{}), NewResourcePermissionsValueNull(), diags
+	}
+
+	// Populate Tenants
+	tenants, ok := cloudDatastore.GetTenantsOk()
+	if !ok || tenants == nil {
+		diags.AddWarning(
+			"populate datastore resource",
+			fmt.Sprintf("datastore %d missing tenants in response", id),
+		)
+
+		return types.SetNull(TenantsType{}), NewResourcePermissionsValueNull(), diags
+	}
+
+	tenantsSet, tdiag := convert.ToSetType(
+		ctx,
+		tenants,
+		func(
+			in sdk.ListCloudDatastores200ResponseAllOfDatastoresInnerTenantsInner,
+		) TenantsValue {
+			return TenantsValue{
+				Id:    convert.Int64ToType(in.Id),
+				state: attr.ValueStateKnown,
+			}
+		},
+	)
+
+	diags = append(diags, tdiag...)
+
+	// Populate ResourcePermissions, we'll only do Groups for now
+	rp, ok := cloudDatastore.GetResourcePermissionOk()
+	if !ok || rp == nil {
+		diags.AddWarning(
+			"populate datastore resource",
+			fmt.Sprintf("datastore %d missing resource permissions in response", id),
+		)
+
+		return tenantsSet, NewResourcePermissionsValueNull(), diags
+	}
+
+	sites, ok := rp.GetSitesOk()
+	if !ok || sites == nil {
+		diags.AddWarning(
+			"populate datastore resource",
+			fmt.Sprintf("datastore %d missing resource permission sites in response", id),
+		)
+
+		return tenantsSet, NewResourcePermissionsValueNull(), diags
+	}
+	groupsSet, gdiag := convert.ToSetType(
+		ctx,
+		sites,
+		func(
+			in sdk.ListCloudDatastores200ResponseAllOfDatastoresInnerResourcePermissionSitesInner,
+		) GroupsValue {
+			return GroupsValue{
+				Id:    convert.Int64ToType(in.Id),
+				state: attr.ValueStateKnown,
+			}
+		},
+	)
+
+	diags = append(diags, gdiag...)
+
+	resourcePermissions, rdiags := populateResourcePermissionsFromApi(ctx, groupsSet)
+	diags = append(diags, rdiags...)
+
+	return tenantsSet, resourcePermissions, diags
+}
+
+// populateClusterDatastoreInformation gets the cluster datastore information for a datastore
+// and populates the Tenants and ResourcePermissions fields of the datastore resource model
+// only to be called on Import
+func populateClusterDatastoreInformation(
+	ctx context.Context,
+	id, clusterId int64,
+	client *sdk.APIClient,
+) (types.Set, ResourcePermissionsValue, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	// Get cluster datastore information
+	cdResp, hresp, err := client.ClustersAPI.GetClusterDatastore(ctx, int32(clusterId), id).Execute() // nolint:gosec
```

Comment:
Not going to hold this PR hostage, but the`int32` cast here is an indicator that the openapi spec needs to be updated